File: js/213_HouseRobberII.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function(nums) {
    var n = nums.length;
    if(n < 2) return n > 0 ? nums[0] : 0;
    return Math.max(robHelper(nums,0,n-2), robHelper(nums,1,n-1));
};

var robHelper = function(nums,b,e){
    var cur = 0, pre = 0;
    for(var i = b; i <= e; ++i){
        var t = Math.max(pre + nums[i], cur);
        pre = cur;
        cur = t;
    }
    return cur;
}

File: cpp/213_HouseRobberII.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.empty()) return 0;
        if(nums.size() < 2) return nums[0];
        return std::max(robHelper(nums,0,nums.size()-2),robHelper(nums,1,nums.size()-1));

    }
private:
    int robHelper(vector<int>& nums, int b, int e){
        int cur = 0, pre = 0;
        for(int i = b; i <= e; ++i){
            int t = max(pre + nums[i], cur);
            pre = cur;
            cur = t;
        }
        return cur;
    }
};